plugin/keyring_vault/vault_curl.cc

```diff
@@ -0,0 +1,177 @@
+#include <algorithm>
+#include "vault_curl.h"
+#include "base64.h"
+
+namespace keyring
+{
+
+static size_t write_response_memory(void *contents, size_t size, size_t nmemb, void *userp)
+{
+  size_t realsize = size * nmemb;
+  std::stringstream *read_data = (std::stringstream *)userp;
+
+  read_data->write((char*)contents, realsize);
+  if (!read_data->good())
+    return 0;
+  return realsize;
+}
+
+std::string Vault_curl::get_error_from_curl(CURLcode curl_code)
+{
+  size_t len = strlen(curl_errbuf);
+  std::stringstream ss;
+  if (curl_code != CURLE_OK)
+  {
+    ss << "Curl returned this error code: " << curl_code;
+    ss << " with error message : ";
+    if(len)
+      ss << curl_errbuf;
+    else
+      ss << curl_easy_strerror(curl_code);
+  }
+  return ss.str();
+}
+
+my_bool Vault_curl::init(Vault_credentials *vault_credentials)
+{
+  curl = curl_easy_init();
+  if (curl == NULL)
+  {
+    logger->log(MY_ERROR_LEVEL, "Could not create CURL session");
+    return TRUE;
+  }
+  this->token_header = "X-Vault-Token:" + (*vault_credentials)["token"];
```

Comment:
operator[] for std::map is not const, so write a helper function which would deal with iterators or choose another container for vault_credentials